models/iaf_wang_2002_exact.cpp

```diff
@@ -0,0 +1,568 @@
+    // enforce setting IntegrationStep to step-t; this is of advantage
+    // for a consistent and efficient integration across subsequent
+    // simulation intervals
+    while ( t < B_.step_ )
+    {
+      const int status = gsl_odeiv_evolve_apply( B_.e_,
+        B_.c_,
+        B_.s_,
+        &B_.sys_,              // system of ODE
+        &t,                    // from t
+        B_.step_,              // to t <= step
+        &B_.integration_step_, // integration step size
+        S_.ode_state_ );       // neuronal state
+
+      if ( status != GSL_SUCCESS )
+      {
+        throw GSLSolverFailure( get_name(), status );
+      }
+    }
+
+    // add incoming spikes
+    S_.ode_state_[ State_::s_AMPA ] += B_.spikes_[ AMPA - 1 ].get_value( lag );
+    S_.ode_state_[ State_::s_GABA ] += B_.spikes_[ GABA - 1 ].get_value( lag );
+
+    for ( size_t i = NMDA - 1; i < B_.spikes_.size(); ++i )
+    // i starts at 2, runs through all NMDA spikes
+    {
+
+      // index which starts at 0
+      const size_t si = i - ( NMDA - 1 );
+
+      assert( si >= 0 );
+      assert( State_::s_NMDA_base + si * 2 <= S_.state_vec_size );
+
+      S_.ode_state_[ State_::s_NMDA_base + si * 2 ] += B_.spikes_.at( i ).get_value( lag );
+    }
+
+    // absolute refractory period
+    if ( S_.r_ )
+    {
+      // neuron is absolute refractory
+      --S_.r_;
+      S_.ode_state_[ State_::V_m ] = P_.V_reset; // clamp potential
+    }
+    else if ( S_.ode_state_[ State_::V_m ] >= P_.V_th )
+    {
+      // neuron is not absolute refractory
+      S_.r_ = V_.RefractoryCounts;
+      S_.ode_state_[ State_::V_m ] = P_.V_reset;
+
+      // log spike with ArchivingNode
+      set_spiketime( Time::step( origin.get_steps() + lag + 1 ) );
+
+      SpikeEvent se;
+      kernel().event_delivery_manager.send( *this, se, lag );
+    }
+
+    // set new input current
+    B_.I_stim_ = B_.currents_.get_value( lag );
+
+    // voltage logging
+    B_.logger_.record_data( origin.get_steps() + lag );
+  }
+}
+
+// Do not move this function as inline to h-file. It depends on
+// universal_data_logger_impl.h being included here.
+void
+nest::iaf_wang_2002_exact::handle( DataLoggingRequest& e )
+{
+  B_.logger_.handle( e );
+}
+
+void
+nest::iaf_wang_2002_exact::handle( SpikeEvent& e )
+{
+  assert( e.get_delay_steps() > 0 );
+  assert( e.get_rport() <= static_cast< int >( B_.spikes_.size() ) );
+
+  const double steps = e.get_rel_delivery_steps( kernel().simulation_manager.get_slice_origin() );
+  const auto rport = e.get_rport();
+
+  if ( rport < NMDA )
+  {
+  B_.spikes_[ rport - 1 ].add_value( steps, e.get_weight() * e.get_multiplicity() );
+  }
+  else
+  // we need to scale each individual S_j variable by its weight,
+  // so we store them
+  {
+    B_.spikes_[ rport - 1 ].add_value( steps, e.get_multiplicity() );
+    const size_t w_idx = rport - NMDA;
+    if ( B_.weights_[ w_idx ] != e.get_weight() )
+    {
+      B_.weights_[ w_idx ] = e.get_weight();
```

Comment:
Throw exception because weight change cannot be implemented properly due to way model is defined.

Reply:
Added